tests/test_traits_core.cc

```diff
@@ -0,0 +1,317 @@
+                                fit_type<HasValidXYFitImpl, X>::type>::value));
+  EXPECT_TRUE(bool(std::is_same<Fit<HasValidXYFitImpl, Y>,
+                                fit_type<HasValidXYFitImpl, Y>::type>::value));
+}
+
+template <typename T>
+struct Base {};
+
+struct Derived : public Base<Derived> {};
+
+TEST(test_traits_core, test_is_valid_fit_type) {
+  EXPECT_TRUE(bool(is_valid_fit_type<Derived, Fit<Derived, X>>::value));
+  EXPECT_TRUE(bool(is_valid_fit_type<Derived, Fit<Derived, Y>>::value));
+  EXPECT_TRUE(bool(is_valid_fit_type<Base<Derived>, Fit<Base<Derived>, X>>::value));
+  EXPECT_TRUE(bool(is_valid_fit_type<Base<Derived>, Fit<Base<Derived>, Y>>::value));
+  // If a Derived class which inherits from Base<Derived> has a fit
+  // which returns Fit<Base<Derived>> consider that a valid fit type.
+  EXPECT_TRUE(bool(is_valid_fit_type<Derived, Fit<Base<Derived>, X>>::value));
+  EXPECT_TRUE(bool(is_valid_fit_type<Derived, Fit<Base<Derived>, Y>>::value));
+  // However a fit_impl which returns
+  EXPECT_FALSE(bool(is_valid_fit_type<Base<Derived>, Fit<Derived, X>>::value));
+  EXPECT_FALSE(bool(is_valid_fit_type<Base<Derived>, Fit<Derived, Y>>::value));
+}
+
+template <typename T>
+struct Adaptable;
+
+template <typename T, typename FeatureType>
+struct Fit<Adaptable<T>, FeatureType> {};
+
+template <typename ImplType>
+struct Adaptable : public ModelBase<Adaptable<ImplType>> {
+
+  Fit<Adaptable<ImplType>, X> fit(const std::vector<X> &,
+                                   const MarginalDistribution &) const {
+    return Fit<Adaptable<ImplType>, X>();
+  }
+
+  /*
+   * This forwards on `fit` definitions from the implementing class
+   * to this class so it can be picked up by ModelBase.
+   */
+  template <typename FeatureType,
+            typename std::enable_if<
+                       has_valid_fit<ImplType, FeatureType>::value,
+                                   int>::type = 0>
+  auto fit(const std::vector<FeatureType> &features,
+                                   const MarginalDistribution &targets) const {
+    return impl().fit(features, targets);
+  }
+
+  /*
+   * CRTP Helpers
+   */
+  ImplType &impl() { return *static_cast<ImplType *>(this); }
+  const ImplType &impl() const { return *static_cast<const ImplType *>(this); }
+
+};
+
+struct Extended : public Adaptable<Extended> {
+
+  using Base = Adaptable<Extended>;
+
+  auto fit(const std::vector<Y> &,
+                 const MarginalDistribution &targets) const {
+    std::vector<X> xs = {{}};
+    return Base::fit(xs, targets);
+  }
+
+  Z predict(const std::vector<Y> &,
+            const Fit<Adaptable<Extended>, X> &,
+            PredictTypeIdentity<Z>) const {
+    return {};
+  }
+
+};
+
+TEST(test_traits_core, test_adaptable_fit_type) {
+  EXPECT_TRUE(bool(std::is_base_of<Fit<Adaptable<Extended>, X>,
+                   fit_type<Extended, Y>::type>::value));
+  EXPECT_TRUE(bool(std::is_base_of<Fit<Adaptable<Extended>, X>,
```

Comment:
Some failure cases would be good here.

Reply:
same as above: I can't set up failure cases for these since inspecting something without a fit_type gives compile error.  I think that's OK though.